ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -523,24 +528,104 @@ KEXT_STATIC int HandleFileOpOperation(
+    {
+        const char* fromPath = reinterpret_cast<const char*>(arg0);
+        const char* newPath = reinterpret_cast<const char*>(arg1);
+        
+        // TODO(Mac): We need to handle failures to lookup the vnode.  If we fail to lookup the vnode
+        // it's possible that we'll miss notifications
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnode, context);
+        if (0 != toErr)
         {
-            PerfSample hardLinkSample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+            KextLog_Error("HandleFileOpOperation: vnode_lookup failed, errno %d for path '%s'", toErr, newPath);
+            goto CleanupAndReturn;
+        }
         
-            int kauthResult;
-            int kauthError;
-            if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
-                    MessageType_KtoU_NotifyFileHardLinkCreated,
-                    currentVnode,
-                    vnodeFsidInode,
-                    newPath,
-                    pid,
-                    procname,
-                    &kauthResult,
-                    &kauthError))
+        // Don't expect named stream here as they can't be directly hardlinked or renamed, only the main fork can
+        assert(!vnode_isnamedstream(currentVnode));
+        
+        putCurrentVnode = true;
+        
+        bool isDirectory = (0 != vnode_isdir(currentVnode));
+        
+        VirtualizationRootHandle targetRoot, fromRoot;
+        pid_t pid;
+        bool messageTargetProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                nullptr, // don't pass path for target provider
+                action,
+                isDirectory,
+                &targetRoot,
+                &pid);
+        bool messageFromProvider =
+            ShouldHandleFileOpEvent(
+                &perfTracer,
+                context,
+                currentVnode,
+                fromPath,
+                action,
+                isDirectory,
+                &fromRoot,
+                &pid);
+        
+        if (!messageTargetProvider && !messageFromProvider)
+        {
+            goto CleanupAndReturn;
+        }
+        
+        FsidInode vnodeFsidInode = Vnode_GetFsidAndInode(currentVnode, context, true /* the inode is used for getting the path in the provider, so use linkid */);
+
+        char procname[MAXCOMLEN + 1];
+        proc_name(pid, procname, MAXCOMLEN + 1);
+
+        {
+            PerfSample sample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+
+            if (messageTargetProvider)
             {
-                goto CleanupAndReturn;
+                int kauthResult;
+                int kauthError = 0;
+                if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
+                        targetRoot,
+                        MessageType_KtoU_NotifyFileHardLinkCreated,
```

Comment:
Related to above question, if this is for a directory do we need a new `MessageType_KtoU_NotifyDirectoryHardLinkCreated` message?

Reply:
@pmj let me know if @wilbaker's comment here makes sense.  i think we want a separation notification message then for directory renames (similar to renames).
I can pick this work up too if you want, not sure if crosses too much into non-kext side work.

Reply:
@wilbaker  this does make me wonder if hard links are allowed on windows directories

Reply:
@jeschu1 Windows does not support directory hardlinks:

https://docs.microsoft.com/en-us/windows/desktop/FileIO/hard-links-and-junctions

I think it makes sense to have a separate PR for directory hardlinks (how we handle them might require a larger design discussion).  

I filed #1152 for directly hardlink support.

Reply:
@pmj let's leave everything as here regarding directories per @wilbaker  suggestion.  We can create a new message type in a seperate PR.